src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -91,57 +92,20 @@ private static IEnumerable<T> Prioritize<T>(IEnumerable<T> items, Func<T, bool>
     /// <code>Parallel.ForEachAsync</code>, allowing for parallel processing of the items, with a preference towards
     /// earlier items.
     /// </summary>
-    private static async Task PerformParallelSearchAsync<T>(
+    private static Task PerformParallelSearchAsync<T>(
         IEnumerable<T> items,
-        Func<T, Action<RoslynNavigateToItem>, CancellationToken, ValueTask> callback,
+        Func<T, Action<RoslynNavigateToItem>, ValueTask> callback,
         Func<ImmutableArray<RoslynNavigateToItem>, Task> onItemsFound,
         CancellationToken cancellationToken)
     {
         // Use an unbounded channel to allow the writing work to write as many items as it can find without blocking.
         // Concurrently, the reading task will grab items when available, and send them over to the host.
         var channel = Channel.CreateUnbounded<RoslynNavigateToItem>(s_channelOptions);
 
-        await Task.WhenAll(
-            FindAllItemsAndWriteToChannelAsync(),
-            ReadItemsFromChannelAndReportToCallbackAsync()).ConfigureAwait(false);
-
-        async Task ReadItemsFromChannelAndReportToCallbackAsync()
-        {
-            await Task.Yield().ConfigureAwait(false);
-            using var _ = ArrayBuilder<RoslynNavigateToItem>.GetInstance(out var items);
-
-            while (await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
-            {
-                // Grab as many items as we can from the channel at once and report in a batch.
-                while (channel.Reader.TryRead(out var item))
-                    items.Add(item);
-
-                await onItemsFound(items.ToImmutableAndClear()).ConfigureAwait(false);
-            }
-        }
-
-        async Task FindAllItemsAndWriteToChannelAsync()
-        {
-            Exception? exception = null;
-            try
-            {
-                await Task.Yield().ConfigureAwait(false);
-                await PerformSearchAsync(item => channel.Writer.TryWrite(item)).ConfigureAwait(false);
-            }
-            catch (Exception ex) when ((exception = ex) == null)
-            {
-                throw ExceptionUtilities.Unreachable();
-            }
-            finally
-            {
-                // No matter what path we take (exceptional or non-exceptional), always complete the channel so the
-                // writing task knows it's done.
-                channel.Writer.TryComplete(exception);
-            }
-        }
-
-        Task PerformSearchAsync(Action<RoslynNavigateToItem> onItemFound)
-            => RoslynParallel.ForEachAsync(
-                items, cancellationToken, (item, cancellationToken) => callback(item, onItemFound, cancellationToken));
+        return channel.RunProducerConsumerAsync(
```

Comment:
> channel.RunProducerConsumerAsync [](http://example.com/codeflow?start=14&length=33)

Does this even need to expose the fact that this producer/consumer mechanism uses channels?

Reply:
I can't think of a good way around that.  Esp. as its the channel that controls backpressure, and things like if it's an Unbounded or Bounded channel, and if you can have multi-production and/or multi-consumption.  

If we don't put this on a channel, we'd need to still have all the channel options.  So it just seemed sane to me that just operate on a channel.

Any ideas?

Reply:
I just thought that this code would pass in the channel options to a general ProducerConsumer class's RunAsync method which handles the channel creation in there. Maybe I'm just not seeing where the fault lies in that.

Reply:
We'd still be makgn the choice on the type of channel to create.  but i suppose that part is ok :)

Reply:
done!

Reply:
This is better and good enough for me. In my head, I had the callers as something like:

struct ProducerConsumerOptions 
{
  bool singleReader;
  bool singleWriter;
}

ProducerConsumer<RoslynNavigateToItem>.RunAsync(
  options: s_producerConsumerOptions, 
  ...);

Reply:
i like that name.